docs/proposals/subscriber-fan-in-fan-out.md

```diff
@@ -0,0 +1,189 @@
+the previous issue, the current pattern means only one subscriber can
+deal with a given message at any one time. This makes it difficult to have a
+separation of concerns.  Ideally, we would be able to have each subscriber
+responsible for doing just one thing (recording metrics adding various
+logs, persisting state etc all based on the msg it had been passed). This
+would make it simple to add more functionality based on these messages
+along with making testing simple and isolated.
+
+
+## Goal
+
+- Outline a design for passing messages, both user-generated and system
+generated from executing APB jobs to each of a set of registered subscribers
+that removes the outlined problems and improves the overall design of the broker.
+
+## Subscriber design principal
+
+This proposal outlines changes to the subscribers and allowing
+the addition of more subscribers allowing them all to act on the same message.
+The order these subscribers are called in should not matter, their work
+should be independent of any other subscribers work.
+
+## Proposal
+
+
+### Change the subscriber interface
+
+Currently, the subscribers implement the WorkSubscriber interface, we
+would change this interface and the current subscribers.
+Rather than passing the channel into the subscriber, we would rename and
+change this interface to accept a ```JobMsg```, value not a pointer to stop
+unexpected mutation, and return an error.
+The current subscribers would handle state persistence and would be
+renamed to reflect their role.
+ 
+```go
+#Note I call it observer here see terminology
+type WorkObserver interface{
+   Notify(msg JobMsg)error
+}
+```
+
+
+#### Change how subscribers are referenced
+
+Currently, when a subscriber is added, we take the existing channel or
+create one if not present and hand that channel to the subscriber.
+Instead, we would now have a registry of subscribers to topics:
+
+```go
+type WorkEngine struct {
+	subscribers map[WorkTopic][]WorkSubscriber
+	...
+} 
+
+```
+
+
+
+### Channel changes
+There a couple of options available as the goal here is to avoid
+any blocking:
+
+#### 1) Keep the single channel but remove potential for blocking
+
+Currently, the topic channel is handed to the single subscriber which
+performs all the work. As mentioned this means no more messages can be
+read from the channel until this unit of work is complete. A change to
+how the messages are handled could alleviate any potential issue. The
+reader would change to be the work engine which would start a read loop
+for each of the topics when the broker started and distribute incoming
+messages to any number of subscribers async:
+As long as the principal that subscribers shouldn't depend on each other
+is maintained there shouldn't be issues with this approach.
+```go 
+for{
+
+   select{
+       case msg := <-provisionChan:
+           for _,s := range e.subscribers[ProvisionTopic]{
+                go s(msg)              
```

Comment:
There's a bit of an inconsistency here I think. (I'm going to use `Subscriber` here for the sake of clarity, assume it is synonymous with `Observer`). If we are moving away from subscriber's taking a channel as an argument with `   Notify(msg JobMsg)error`, this should be `s.Notify(msg)`. Additionally, this is fire-and-forget; the engine doesn't have an opportunity to deal with errors that may come back. That may be desirable for this soln? The question I would ask is:

* Is it okay to just fire these off and ignore the outcome? If so, it doesn't make sense to return an error from `Notify`. How do we handle error scenarios in the subscribers then, if at all?
* If it's not okay, how do we expect the engine to handle errors that have been returned from subscribers?

Reply:
👍 I should not have had the error return on the interface, I will  change this. As each WorkSubscriber is an independent function unaware of any of the other work that is happening, I think it makes sense for them to handle the error internally (similar to how the subscribers work now, marking the action as failed, cleaning up etc) as the only real options for the WorkEngine  on receiving an error would be to log it.